### ISSUE TITLE
Metabase doesn't work with Postgres 10 beta

### ISSUE THREAD

**Issue by alitrack** [metabase/metabase#5424]
- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, IE 10)
Chrome 52.1
- Your operating system: (e.x. OS X 10, Windows XP, etc)
Windows 2008R2
- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
PostgreSQL 10 beta1 Windows version
- Metabase version: (e.x. 0.19.3)
0.24.2
- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
Windows 
- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
H2
- *Repeatable steps to reproduce the issue*
use PostgreSQL 10 beta 1 will always get error, 

```
七月 05 18:48:33 INFO metabase.sync-database.introspect :: Introspecting schema on PostgreSQL database 'arules' ...
七月 05 18:49:02 ERROR metabase.events.sync-database :: Error syncing Database: 6 #error {
:cause A ResourcePool could not acquire a resource from its primary factory or source.
:via
[{:type java.sql.SQLException
:message Connections could not be acquired from the underlying database!
:at [com.mchange.v2.sql.SqlUtils toSQLException SqlUtils.java 118]}
{:type com.mchange.v2.resourcepool.CannotAcquireResourceException
:message A ResourcePool could not acquire a resource from its primary factory or source.
:at [com.mchange.v2.resourcepool.BasicResourcePool awaitAvailable BasicResourcePool.java 1319]}]
:trace
[[com.mchange.v2.resourcepool.BasicResourcePool awaitAvailable BasicResourcePool.java 1319]
[com.mchange.v2.resourcepool.BasicResourcePool prelimCheckoutResource BasicResourcePool.java 557]
[com.mchange.v2.resourcepool.BasicResourcePool checkoutResource BasicResourcePool.java 477]
[com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool checkoutPooledConnection C3P0PooledConnectionPool.java 525]
[com.mchange.v2.c3p0.impl.AbstractPoolBackedDataSource getConnection AbstractPoolBackedDataSource.java 128]
[clojure.java.jdbc$get_connection invokeStatic jdbc.clj 298]
[clojure.java.jdbc$get_connection invoke jdbc.clj 193]
[metabase.driver.generic_sql$describe_database invokeStatic generic_sql.clj 410]
[metabase.driver.generic_sql$describe_database invoke generic_sql.clj 407]
[metabase.driver$fn__22369$G__22335__22376 invoke driver.clj 45]
[metabase.sync_database.introspect$db__GT_tables invokeStatic introspect.clj 188]
[metabase.sync_database.introspect$db__GT_tables invoke introspect.clj 187]
[metabase.sync_database.introspect$introspect_database_and_update_raw_tables_BANG_ invokeStatic introspect.clj 204]
[metabase.sync_database.introspect$introspect_database_and_update_raw_tables_BANG_ invoke introspect.clj 198]
[metabase.sync_database$sync_database_with_tracking_BANG_$fn__29713 invoke sync_database.clj 78]
[metabase.sync_database$sync_database_with_tracking_BANG_ invokeStatic sync_database.clj 75]
[metabase.sync_database$sync_database_with_tracking_BANG_ invoke sync_database.clj 69]
[clojure.core$partial$fn__4763 invoke core.clj 2528]
[metabase.driver$fn__22593 invokeStatic driver.clj 251]
[metabase.driver$fn__22593 invoke driver.clj 251]
[metabase.driver$fn__22542$G__22325__22551 invoke driver.clj 45]
[metabase.sync_database$sync_database_BANG_ invokeStatic sync_database.clj 44]
[metabase.sync_database$sync_database_BANG_ doInvoke sync_database.clj 25]
[clojure.lang.RestFn invoke RestFn.java 410]
[metabase.events.sync_database$process_sync_database_event$fn__44728 invoke sync_database.clj 31]
[clojure.core$binding_conveyor_fn$fn__4676 invoke core.clj 1938]
[clojure.lang.AFn call AFn.java 18]
[java.util.concurrent.FutureTask run nil -1]
[java.util.concurrent.ThreadPoolExecutor runWorker nil -1]
[java.util.concurrent.ThreadPoolExecutor$Worker run nil -1]
[java.lang.Thread run nil -1]]}
```

**Comment by salsakran** [comment]
are you running PG on windows as well?

**Comment by witwall** [comment]
works well with other client, for example pgAdmin4, DBeaver.
If connect with metabase, all connection will be used, I tried set max_connection to 1000, same issue.

**Comment by camsaul** [comment]
maybe the Postgres JDBC driver or something else we're using doesn't like the Postgres 10 beta? @witwall does it work with Postgres 9?

**Comment by witwall** [comment]
I tested Postgres 9.5 and 9.6.x, works.

**Comment by camsaul** [comment]
@salsakran it sounds like the Postgres JDBC driver we're using isn't working yet with the new Postgres 10 beta, so this isn't something we can really fix on our end. Is the protocol the same for this as it is for the OpenJDK beta bugs we've seen? Close as "Won't Fix" since it's an early-stage beta?

**Comment by salsakran** [comment]
PG 10 is supposed to be out sometime in Q3 2017 (https://www.postgresql.org/developer/roadmap/), which is pretty short term. 

I'd say this is currently a P3, and will be elevated each month. Once the PG 10 release date comes we should support it fully.

**Comment by camsaul** [comment]
I'm sure we will indeed support it fully but that's more of a function of when the PG people update their JDBC driver (which I'm sure they will do sooner or later). Does it make sense to keep open an inactionable bug to track something upstream like this?

**Comment by salsakran** [comment]
We should always keep open bugs like this that track user perceived bugs that are a result of upstream issues. This is both a reminder that there is ShitThat'sBrokenInTheApp, and so that if someone searches for the issue, they see its status vs opening another issue.

**Comment by witwall** [comment]
I found the solution, use a newer  version postgres jdbc, for example, 
postgresql-9.4.1212 or postgresql-42.1.3, I tried both.

```
java -cp postgresql-9.4.1212.jre7.jar:metabase.jar metabase.core
```

btw, I tried it under Windows, it works too.

```
java -cp postgresql-42.1.3.jre7.jar;metabase.jar metabase.core
```

so It can be closed.

**Comment by witwall** [comment]
@camsaul @salsakran 
please have a look the solution.

**Comment by camsaul** [comment]
@witwall good catch. I didn't realize our Postgres JDBC driver was out-of-date. I'm checked our dependencies with [`lein-ancient`](https://github.com/xsc/lein-ancient) but Postgres didn't show up as out-of-date. I'm guessing it's because we're still on a 9.3.x driver and Postgres recently had that [huge version bump where they went from 9.x to 42.0](https://jdbc.postgresql.org/documentation/faq.html#version-change). Anyways I'll update our dependency so you won't need to manually add the updated JDBC driver to the classpath in our upcoming release

**Comment by camsaul** [comment]
Fixed by #5511

**Comment by salsakran** [comment]
Unfortunately, the new driver doesn't work with Redshift. Reverting this change and re-opening this issue until we can figure out a way to upgrade and support PG 10 without breaking Redshift support. Expect this in 0.25.1

**Comment by witwall** [comment]
@salsakran maybe we should use Redshift Redshift, not PostgreSQL Redshift

http://docs.aws.amazon.com/redshift/latest/mgmt/configure-jdbc-connection.html#download-jdbc-driver
or try postgresql-42.1.3.jre7.jar

btw, why not use JDBC 4.2 API?

**Comment by salsakran** [comment]
https://github.com/metabase/metabase/pull/5511/ tried the `postgresql-42.1.3.jre7.jar` driver and it barfs on redshift.

Based on my read of the license for the redshift jdbc driver we can't distribute it with GPL software. 

Not sure what the best course of action is, but we're trying a couple of things out and have reached out to AWS to clarify the driver distribution situation.

**Comment by witwall** [comment]
@salsakran sorry my typo, I want to say have a try postgresql-9.4.1212.jre7.jar

**Comment by camsaul** [comment]
@salsakran have you heard back from the Amazon people WRT our ability to distribute the redshift JDBC driver?

**Comment by jnicholls** [comment]
Any news?  Metabase is going to be in a world of hurt when PG 10 is released and it's still not functioning.  I'd start driver hacking quickly.

**Comment by astathis** [comment]
With postgres 10 beta 4, I can't still get connection. For example, if I make a new database connection I don't see any table on any schema but with other tools like pgAdmin everything is working fine

**Comment by jnicholls** [comment]
What I did was I updated the postgres driver to the latest version and rebuilt, and everything is great now.

**Comment by witwall** [comment]
@jnicholls I gave the solution in this thread without compile it.

**Comment by astathis** [comment]
@jnicholls Can you write the command that you execute Metabase? It's something like that : 
            `java -cp postgresql-9.4.1212.jre7.jar:metabase.jar metabase.core`

**Comment by senior** [comment]
@trifle The PR you're referring to is a few months old. It was necessary to revert that as it broke Redshift and we needed to figure out a way to support both. Last week I merged in the update  https://github.com/metabase/metabase/commit/d23b26694056456cc370580841820dc4f55ae9a4 to both Redshift and PostgreSQL. It's not been released, but is currently on master.

**Comment by trifle** [comment]
Thanks, that is very good to know! I could only find this issue and the reverted commit, hence my comment - but now there's a trail for others to follow :)

By the way, thanks for all your work on metabase - it's an incredible tool!

**Comment by senior** [comment]
Note that the fix for this was included with 0.26.2 which was just released. Upgrading should fix this issue.
